superset/sql_parse.py

```diff
@@ -206,14 +276,91 @@ def __init__(self, sql_statement: str, strip_comments: bool = False):
+        """
+        Extract all table references in a query.
+
+        Note: this uses sqlglot, since it's better at catching more edge cases.
+        """
+        try:
+            statements = parse(self.sql, dialect=self._dialect)
+        except ParseError:
+            logger.warning("Unable to parse SQL (%s): %s", self._dialect, self.sql)
+            return set()
+
+        return {
+            table
+            for statement in statements
+            for table in self._extract_tables_from_statement(statement)
+            if statement
+        }
+
+    def _extract_tables_from_statement(self, statement: exp.Expression) -> set[Table]:
+        """
+        Extract all table references in a single statement.
+
+        Please not that this is not trivial; consider the following queries:
+
+            DESCRIBE some_table;
+            SHOW PARTITIONS FROM some_table;
+            WITH masked_name AS (SELECT * FROM some_table) SELECT * FROM masked_name;
+
+        See the unit tests for other tricky cases.
+        """
+        sources: Iterable[exp.Table]
+
+        if isinstance(statement, exp.Describe):
+            # A `DESCRIBE` query has no sources in sqlglot, so we need to explicitly
+            # query for all tables.
+            sources = statement.find_all(exp.Table)
+        elif isinstance(statement, exp.Command):
+            # Commands, like `SHOW COLUMNS FROM foo`, have to be converted into a
+            # `SELECT` statetement in order to extract tables.
+            literal = statement.find(exp.Literal)
+            if not literal:
+                return set()
+
+            pseudo_query = parse_one(f"SELECT {literal.this}", dialect=self._dialect)
+            sources = pseudo_query.find_all(exp.Table)
+        elif statement:
+            sources = []
+            for scope in traverse_scope(statement):
+                for source in scope.sources.values():
+                    if not isinstance(source, exp.Table):
+                        continue
+
+                    # CTEs in the parent scope look like tables (and are represented by
+                    # exp.Table objects), but should not be considered as such;
+                    # otherwise a user with access to table `foo` could access any table
+                    # with a query like this:
+                    #
+                    #   WITH foo AS (SELECT * FROM bar) SELECT * FROM foo
+                    #
+                    parent_sources = scope.parent.sources if scope.parent else {}
+                    ctes_in_scope = {
+                        name
+                        for name, parent_scope in parent_sources.items()
+                        if isinstance(parent_scope, Scope)
+                        and parent_scope.scope_type == ScopeType.CTE
+                    }
+                    if source.name not in ctes_in_scope:
+                        sources.append(source)
+
+        else:
```

Comment:
Does this scenario just represent incorrect use of the method? Or is there some scenario in which an `exp.Expression` could be falsy, like an empty query?

I see the codecov plugin has complained it's not covered by your tests, so I guess trying to write a test for this scenario may answer whether it needs special handling.

Reply:
Ah, this is leftover code. In sqlglot the result of parsing SQL is `Optional[Expression]`, so I added this logic. But we now take care of false value sin the caller.